packages/labs/src/controllers/context.ts

```diff
@@ -0,0 +1,324 @@
+        consumer.value = this.value;
+      }
+    }
+  }
+}
+
+const pendingConsumers: WeakMap<Element, Consumer[]> = new WeakMap();
+
+export class Consumer extends UpdatingController {
+  strategy = DEFAULT_STRATEGY;
+
+  // @internal
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  _value: any = undefined;
+
+  provider?: Provider;
+  key = null;
+
+  /**
+   * Sends connection signal to provider. The first ancestor provider with
+   * the appropriate key will become this consumer's provider.
+   */
+  connectedCallback() {
+    // Purely as an optimization, batch connection to update if one is pending.
+    // This allows all consumers for the element to be connected with one
+    // provider search.
+    if (this.host.isUpdatePending && !this.host.hasUpdated) {
+      let pending = pendingConsumers.get(this.host);
+      if (pending === undefined) {
+        pendingConsumers.set(this.host, (pending = []));
+      }
+      pending.push(this);
+    } else {
+      this._connectToProvider();
+    }
+  }
+
+  // Batches connection to update if possible
+  willUpdate() {
+    if (!this.host.hasUpdated && this.provider === undefined) {
+      const pending = pendingConsumers.get(this.host);
+      if (pending?.length) {
+        this._connectToProvider(pending);
+        // pending consumers get one shot to find a provider.
+        pending.length = 0;
+      }
+    }
+  }
+
+  // @internal
+  _connectToProvider(consumers: Consumer[] = [this]) {
+    // TODO(sorvell): Multiple strategies are implemented primarily
+    // to test performance.
+    // Find provider by firing event.
+    if (this.strategy === EVENT_STRATEGY) {
+      this.host.dispatchEvent(
+        new CustomEvent(connectContextEvent, {
+          detail: consumers,
+          composed: true,
+          bubbles: true,
+        })
+      );
+      // Or ascend tree to find provider.
+    } else {
+      let node: Node = this.host;
+      let connectedCount = 0;
+      while ((node = node.parentNode ?? (node as ShadowRoot).host)) {
+        const providers = providerMap.get(node);
+        if (providers !== undefined) {
+          // Try to find a provider for the consumers.
+          for (let i = 0; i < consumers.length; i++) {
+            const consumer = consumers[i];
+            const provider = providers.find(
+              (provider) => provider.key === consumer.key
+            );
+            // If a provider is found, connect the consumer and remove it from
+            // our search list.
+            if (provider !== undefined) {
+              provider.connect(consumer);
+              connectedCount++;
+            }
+          }
+        }
+        if (connectedCount === consumers.length) {
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Disconnects from the provider.
+   */
+  disconnectedCallback() {
+    this.provider?.disconnect(this);
+  }
+
+  get value() {
+    return this._value;
+  }
+
+  set value(value) {
```

Comment:
It could even be private-ish, since it's not intended for use by the `Consumer` user, but by the system; so that the `value` property is read-only from user POV.

Reply:
Done.